File: scripts/generate_event_pages.py
#!/usr/bin/env python3
import json
import os
import sys
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Tuple

ROOT = Path(__file__).resolve().parent.parent
INPUT_DIR = ROOT / "events-data"
OUTPUT_DIR = ROOT / "app" / "events"
SITEMAP_PATH = ROOT / "public" / "sitemap.xml"
LOG_PATH = ROOT / "event-generation.log"

REQUIRED_FIELDS = ["source_url", "event_date", "status"]


def fail(message: str) -> None:
    print(f"ERROR: {message}", file=sys.stderr)
    raise SystemExit(1)


def load_event_files(input_dir: Path) -> List[Path]:
    if not input_dir.exists():
        fail(f"Input folder not found: {input_dir}")
    files = sorted(input_dir.glob("*.json"))
    if not files:
        fail(f"No JSON event files found in {input_dir}")
    return files


def validate_event_payload(payload: Dict[str, Any], path: Path) -> Tuple[bool, str | None]:
    for field in REQUIRED_FIELDS:
        value = payload.get(field)
        if value in (None, "", []):
            return False, f"missing required field '{field}'"

    if payload.get("status") not in {"confirmed", "estimated", "draft"}:
        return False, "status must be one of: confirmed, estimated, draft"

    if payload.get("source_url") and not isinstance(payload.get("source_url"), str):
        return False, "source_url must be a string"

    return True, None


def slugify(value: str) -> str:
    text = value.strip().lower()
    text = ''.join(ch if ch.isalnum() else '-' for ch in text)
    text = '-'.join(part for part in text.split('-') if part)
    return text or "event"


def generate_page_component(payload: Dict[str, Any], slug: str) -> str:
    data = json.dumps(payload, indent=2)
    return f"""import EventDetailPage, {{ type EventDetailData }} from '@/components/event-detail-page'

const eventData: EventDetailData = {data}

export default function EventPage() {{
  return <EventDetailPage data={{eventData}} slug=\"{slug}\" />
}}
"""


def write_page(payload: Dict[str, Any], source_path: Path, output_dir: Path) -> Tuple[str, str]:
    slug = slugify(str(payload.get('event_name') or source_path.stem))
    page_dir = output_dir / slug
    page_dir.mkdir(parents=True, exist_ok=True)
    page_path = page_dir / "page.tsx"
    page_path.write_text(generate_page_component(payload, slug), encoding="utf-8")
    return str(page_path.relative_to(ROOT)), slug


def write_sitemap(urls: List[str], output_path: Path) -> None:
    content = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
    content += "<urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\">\n"
    for url in urls:
        content += "  <url>\n"
        content += f"    <loc>{url}</loc>\n"
        content += "  </url>\n"
    content += "</urlset>\n"
    output_path.write_text(content, encoding="utf-8")


def write_summary_log(summary: List[Dict[str, Any]], log_path: Path) -> None:
    lines = ["Event generation summary:"]
    for item in summary:
        lines.append(f"- {item['status']}: {item['file']} -> {item['message']}")
    log_path.write_text("\n".join(lines) + "\n", encoding="utf-8")


def main() -> None:
    input_dir = INPUT_DIR
    output_dir = OUTPUT_DIR
    output_dir.mkdir(parents=True, exist_ok=True)

    event_files = load_event_files(input_dir)
    generated_urls: List[str] = []
    summary: List[Dict[str, Any]] = []
    generated_count = 0
    skipped_count = 0

    for event_file in event_files:
        try:
            with event_file.open("r", encoding="utf-8") as handle:
                payload = json.load(handle)
        except json.JSONDecodeError as exc:
            skipped_count += 1
            summary.append({"status": "skipped", "file": event_file.name, "message": f"invalid JSON: {exc}"})
            continue

        if not isinstance(payload, dict):
            skipped_count += 1
            summary.append({"status": "skipped", "file": event_file.name, "message": "top-level JSON value must be an object"})
            continue

        is_valid, reason = validate_event_payload(payload, event_file)
        if not is_valid:
            skipped_count += 1
            summary.append({"status": "skipped", "file": event_file.name, "message": reason or "validation failed"})
            continue

        page_relative_path, slug = write_page(payload, event_file, output_dir)
        generated_count += 1
        generated_urls.append(f"https://example.com/events/{slug}")
        summary.append({"status": "generated", "file": event_file.name, "message": f"created {page_relative_path}"})

    write_sitemap(generated_urls, SITEMAP_PATH)
    write_summary_log(summary, LOG_PATH)

    print(f"Generated pages: {generated_count}")
    print(f"Skipped pages: {skipped_count}")
    print("Summary log written to", LOG_PATH)
    print("Sitemap written to", SITEMAP_PATH)


if __name__ == "__main__":
    main()

File: scripts/generate-event-pages.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath } from 'node:url';

const scriptDir = path.dirname(fileURLToPath(import.meta.url));
const rootDir = path.resolve(scriptDir, '..');
const inputDir = path.join(rootDir, 'events-data');
const outputDir = path.join(rootDir, 'app', 'events');
const sitemapPath = path.join(rootDir, 'public', 'sitemap.xml');
const logPath = path.join(rootDir, 'event-generation.log');

const sitemapStaticUrls = [
  'https://finvents.vercel.app/',
  'https://finvents.vercel.app/about',
  'https://finvents.vercel.app/contact',
  'https://finvents.vercel.app/privacy',
  'https://finvents.vercel.app/disclaimer',
];

const canonicalEventSlugs = [
  'cpi-release',
  'federal-tax-filing-deadline',
  'fomc-interest-rate-decision-fomc-meeting-2026-09',
  'fomc-interest-rate-decision-fomc-meeting-2026-12',
  'gdp-release',
  'jobs-report-non-farm-payrolls',
];

function slugify(value) {
  return value
    .toString()
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '') || 'event';
}

function toPageContent(payload, slug) {
  const data = JSON.stringify(payload, null, 2);
  return `import EventDetailPage, { type EventDetailData } from '@/components/event-detail-page'

const eventData: EventDetailData = ${data}

export default function EventPage() {
  return <EventDetailPage data={eventData} slug="${slug}" />
}
`;
}

function validatePayload(payload) {
  const missing = ['source_url', 'event_date', 'status'].filter((field) => {
    const value = payload[field];
    return value === undefined || value === null || value === '';
  });

  if (missing.length) {
    return { valid: false, reason: `missing required field(s): ${missing.join(', ')}` };
  }

  if (typeof payload.source_url !== 'string' || payload.source_url.trim() === '' || payload.source_url === '#') {
    return { valid: false, reason: 'source_url must be a real non-dummy URL' };
  }

  if (!['confirmed', 'estimated', 'draft'].includes(payload.status)) {
    return { valid: false, reason: 'status must be one of confirmed, estimated, draft' };
  }

  return { valid: true, reason: null };
}

async function ensureDirectory(dir) {
  await fs.mkdir(dir, { recursive: true });
}

async function writeSitemap(urls) {
  const content = `<?xml version="1.0" encoding="UTF-8"?>\n<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n${urls.map((url) => `  <url>\n    <loc>${url}</loc>\n  </url>`).join('\n')}\n</urlset>\n`;
  await fs.writeFile(sitemapPath, content, 'utf8');
}

async function writeSummaryLog(entries) {
  const lines = ['Event generation summary:'];
  for (const entry of entries) {
    lines.push(`- ${entry.status}: ${entry.file} -> ${entry.message}`);
  }
  await fs.writeFile(logPath, `${lines.join('\n')}\n`, 'utf8');
}

async function main() {
  await ensureDirectory(outputDir);
  await ensureDirectory(path.dirname(sitemapPath));

  const files = (await fs.readdir(inputDir)).filter((file) => file.endsWith('.json')).sort();
  if (!files.length) {
    throw new Error(`No JSON event files found in ${path.relative(rootDir, inputDir)}`);
  }

  const generatedUrls = [];
  const summary = [];
  let generatedCount = 0;
  let skippedCount = 0;

  for (const file of files) {
    const filePath = path.join(inputDir, file);
    try {
      const raw = await fs.readFile(filePath, 'utf8');
      const payload = JSON.parse(raw);
      if (!payload || typeof payload !== 'object' || Array.isArray(payload)) {
        throw new Error('top-level JSON value must be an object');
      }

      const validation = validatePayload(payload);
      if (!validation.valid) {
        skippedCount += 1;
        summary.push({ status: 'skipped', file, message: validation.reason });
        continue;
      }

      if (payload.status === 'confirmed' && (!payload.source_url || payload.source_url === '#')) {
        throw new Error('confirmed events must include a real source_url');
      }

      const baseSlug = slugify(payload.event_name || path.basename(file, '.json'));
      const slug = baseSlug === 'fomc-interest-rate-decision' ? `${baseSlug}-${path.basename(file, '.json')}` : baseSlug;
      const pageDir = path.join(outputDir, slug);
      await ensureDirectory(pageDir);
      await fs.writeFile(path.join(pageDir, 'page.tsx'), toPageContent(payload, slug), 'utf8');

      generatedCount += 1;
      generatedUrls.push(`https://finvents.vercel.app/events/${slug}`);
      summary.push({ status: 'generated', file, message: `created app/events/${slug}/page.tsx` });
    } catch (error) {
      skippedCount += 1;
      summary.push({ status: 'skipped', file, message: error.message });
    }
  }

  const sitemapUrls = [
    ...sitemapStaticUrls,
    ...canonicalEventSlugs.map((slug) => `https://finvents.vercel.app/events/${slug}`),
  ];

  await writeSitemap(sitemapUrls);
  await writeSummaryLog(summary);

  console.log(`Generated pages: ${generatedCount}`);
  console.log(`Skipped pages: ${skippedCount}`);
  console.log(`Summary log written to ${path.relative(rootDir, logPath)}`);
  console.log(`Sitemap written to ${path.relative(rootDir, sitemapPath)}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
